compiler/resolution/caches.cpp

```diff
@@ -19,181 +19,63 @@
  */
 
 #include "caches.h"
-
-#include "astutil.h"
 #include "stmt.h"
-#include "stringutil.h"
-
-
-/************************************* | **************************************
-*                                                                             *
-*                                                                             *
-*                                                                             *
-************************************** | *************************************/
+#include <algorithm>
 
 SymbolMapCache genericsCache;
 SymbolMapCache promotionsCache;
 
-static bool isCacheEntryMatch(SymbolMap* s1, SymbolMap* s2);
-
-SymbolMapCacheEntry::SymbolMapCacheEntry(FnSymbol* ifn, SymbolMap* imap) :
-  fn(ifn), map(*imap) { }
-
+typedef std::pair<int,int> IntPair;
+typedef std::vector<IntPair> PairVec;
 
-void
-addCache(SymbolMapCache& cache,
-         FnSymbol*       oldFn,
-         FnSymbol*       fn,
-         SymbolMap*      map) {
-  Vec<SymbolMapCacheEntry*>* entries = cache.get(oldFn);
-  SymbolMapCacheEntry*       entry   = new SymbolMapCacheEntry(fn, map);
-
-  if (entries) {
-    entries->add(entry);
-
-  } else {
-    entries = new Vec<SymbolMapCacheEntry*>();
-    entries->add(entry);
-    cache.put(oldFn, entries);
-  }
+static bool PairCompare(IntPair p1, IntPair p2) {
+  return p1.first < p2.first;
 }
 
-
-FnSymbol*
-checkCache(SymbolMapCache& cache, FnSymbol* oldFn, SymbolMap* map) {
-  if (Vec<SymbolMapCacheEntry*>* entries = cache.get(oldFn)) {
-    forv_Vec(SymbolMapCacheEntry, entry, *entries) {
-      if (isCacheEntryMatch(map, &entry->map))
-        return entry->fn;
+FnSymbol*& lookupCache(SymbolMapCache& cache, FnSymbol* fn, SymbolMap* map) {
+/*
+First, build the key, which is the pair (fn, map). It is represented
+as a sequence of node IDs for the following nodes
+
+key[0] fn
+key[1] the 1st key in the map
+key[2] the 1st value in the map
+...
+key[2*n-1] the last key in the map
+key[2*n]   the last value in the map
```

Comment:
It seems like this would be better as a type that stores two fields:
- the function
- a map of key to value pairs

This odd/even strategy feels archaic to me, like writing a for loop by hand.  We have better tools now, avoiding using them will just make the code more difficult to maintain.

Reply:
It also means you're doing an additional copy at the end that you don't need, right?

Reply:
A map is what we have there now. My motivation is that comparing two maps is slower than comparing two vectors element-wise.

As a compromise, I can switch the odd/even strategy to a vector of pairs.

What do you think?

Reply:
Doesn't your PR message say you saw little to no performance improvements as a result of switching?  That seems to argue that using a vector instead isn't worthwhile to me . . .

You'll also be able to use the map's find method instead of writing the check by hand in that case, right?